pyperformance/data-files/benchmarks/bm_async_tree/run_benchmark.py

```diff
@@ -0,0 +1,143 @@
+"""
+Benchmark for async tree workload, which calls asyncio.gather() on a tree
+(6 levels deep, 6 branches per level) with the leaf nodes simulating some
+(potentially) async work (depending on the benchmark variant). Benchmark
+variants include:
+
+1) "none": No actual async work in the async tree.
+2) "io": All leaf nodes simulate async IO workload (async sleep 50ms).
+3) "memoization": All leaf nodes simulate async IO workload with 90% of 
+                  the data memoized
+4) "cpu_io_mixed": Half of the leaf nodes simulate CPU-bound workload and 
+                   the other half simulate the same workload as the 
+                   "memoization" variant.
+"""
+
+
+import asyncio
+import math
+import random
+
+import pyperf
+
+
+NUM_RECURSE_LEVELS = 6
+NUM_RECURSE_BRANCHES = 6
+RANDOM_SEED = 0
+IO_SLEEP_TIME = 0.05
+MEMOIZABLE_PERCENTAGE = 90
+CPU_PROBABILITY = 0.5
+FACTORIAL_N = 500
+
+
+class AsyncTree:
+    def __init__(self):
+        self.cache = {}
+        # set to deterministic random, so that the results are reproducible
+        random.seed(RANDOM_SEED)
+
+    async def mock_io_call(self):
+        await asyncio.sleep(IO_SLEEP_TIME)
+
+    async def workload_func(self):
+        raise NotImplementedError(
+            "To be implemented by each variant's derived class."
+        )
+
+    async def recurse(self, recurse_level):
+        if recurse_level == 0:
+            await self.workload_func()
+            return
+
+        await asyncio.gather(
+            *[self.recurse(recurse_level - 1) for _ in range(NUM_RECURSE_BRANCHES)]
+        )
+
+    async def run(self):
+        await self.recurse(NUM_RECURSE_LEVELS)
+
+
+class NoneAsyncTree(AsyncTree):
+    async def workload_func(self):
+        return
+
+
+class IOAsyncTree(AsyncTree):
+    async def workload_func(self):
+        await self.mock_io_call()
+
+
+class MemoizationAsyncTree(AsyncTree):
+    async def workload_func(self):
+        data = random.randint(1, 100)
+
+        if data <= MEMOIZABLE_PERCENTAGE:
+            if self.cache.get(data):
+                return data
+
+            self.cache[data] = True
+
+        await self.mock_io_call()
+        return data
+
+
+class CpuIoMixedAsyncTree(MemoizationAsyncTree):
+    async def workload_func(self):
+        if random.random() < CPU_PROBABILITY:
```

Comment:
Won't this lead to inconsistent results between runs?

Reply:
The random's seed is pre-set on line 37 in `def __init__()`:

```
# set to deterministic random, so that the results are reproducible
random.seed(RANDOM_SEED)
```

So it shouldn't be a problem.

Reply:
Ah, I missed that.  Consider adding a comment here referring to `AsyncTree.__init__()`.